tests/src/Kernel/Entity/GetUserGroupsTest.php

```diff
@@ -234,6 +228,32 @@ public function testIsMemberStates() {
     $this->assertFalse(Og::isMemberPending($this->group1, $this->user3));
   }
 
+  /**
+   * Tests retrieval of groups filtered by roles.
+   */
+  public function testGetGroupsByRoles() {
+    /** @var \Drupal\og\MembershipManagerInterface $membership_manager */
+    $membership_manager = \Drupal::service('og.membership_manager');
+
+    $og_role = OgRole::create();
+    $og_role
+      ->setName($this->randomMachineName())
+      ->setLabel(mb_strtolower($this->randomString()))
+      ->setGroupType('entity_test')
+      ->setGroupBundle($this->groupBundle)
+      ->save();
+
+    $member_role = OgRole::getRole('entity_test', $this->groupBundle, OgRoleInterface::AUTHENTICATED);
+
+    $this->createOgMembership($this->group2, $this->user1, [$og_role->getName()]);
+    $groups = $membership_manager->getUserGroupsByRoles($this->user1, [$member_role]);
+    $this->assertCount(2, $groups['entity_test']);
+    $groups = $membership_manager->getUserGroupsByRoles($this->user1, [$og_role]);
+    $this->assertCount(1, $groups['entity_test']);
```

Comment:
It's only testing basic use cases. The `$states` filter is not tested, and neither is passing multiple roles in the `$roles` array.